<a href="https://colab.research.google.com/github/mahmud-nobe/CS110_pcw/blob/master/CS110_Pre_Class_Work_8_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [0]:
NAME = "Md Mahmudunnobe"
COLLABORATORS = ""

---

In [1]:
!pip install jupyter_contrib_nbextensions


     |████████████████████████████████| 20.9MB 220kB/s 
     |████████████████████████████████| 870kB 49.6MB/s 
     |████████████████████████████████| 481kB 48.1MB/s 
  Created wheel for jupyter-latex-envs: filename=jupyter_latex_envs-1.4.6-py2.py3-none-any.whl size=963395 sha256=a1c9ccd3da2ae87931b18628553115b36881897c252f93548ea57ebdbad02506
  Stored in directory: /root/.cache/pip/wheels/0d/71/2a/164491997299b9f2479a251e254323fe35d946779e18f27956
  Created wheel for jupyter-nbextensions-configurator: filename=jupyter_nbextensions_configurator-0.4.1-py2.py3-none-any.whl size=465825 sha256=4eda304e3764e339db3cb8449a34925e08fc5a9112ef5acd43ea7091ede368f4
  Stored in directory: /root/.cache/pip/wheels/15/df/fe/2a74fe34709e7fdc5ae153a768675d9fda93cc7d5133ed1fb0
Successfully built jupyter-latex-envs jupyter-nbextensions-configurator


In [7]:
jupyter contrib nbextension install --user
jupyter nbextension enable spellchecker/main

NameError: ignored

I mispellde it. 



# CS110 Pre-class Work 8.2

# Question 1. (Exercise 12.2-1, Cormen et al.)

Suppose that we have numbers between 1 and 1000 in a binary search tree, and we want to search for the number 363. Which of the following sequences could not be the sequence of nodes examined?

a. 2, 252, 401, 398, 330, 344, 397, 363.

b. 924, 220, 911, 244, 898, 258, 362, 363.

c. 925, 202, 911, 240, 912, 245, 363.

d. 2, 399, 387, 219, 266, 382, 381, 278, 363.

e. 935, 278, 347, 621, 299, 392, 358, 363.


**(c):**
- 925 > 363, 

## Question 2. Comparing complexities
Complete the following table with the average vs worst case complexities for the data structures in each row.

You should copy the following table and paste and edit it in the cell below. 

Operations | BST | Hash table using open addressing | Min heap
--- | --- | --- | ---
Search | O(h) | O(1) | O(n)
Find max | O(h) | O(n) | O(n)
Find min | O(h) | O(n) | O(1)
Max extraction  | O(h) | O(n) | O(n)
Min extraction | O(h) | O(n) | O(1)
Find successor | O(h) | - | -
Find predecessor | O(h) | - | -
Insert | O(h) | O(1) | O(lg n)
Delete | O(h) | O(1) | O(lg n)



YOUR ANSWER HERE

## Question 3. Programming a recursive BST
The code given in the cell below, write python code for the corresponding functions:

* function `search(self, value)`: searches a *non-empty* BST rooted at the node for a node with `data=value`, returns the node if found, None otherwise
* function `delete(self, value)`: if a node with data = value is present in the tree rooted at Node, deletes that node and returns the root.
* function `inorder(self)`: returns a list of all data in the tree rooted at root produced using an in order traversal. When correctly implemented on a BST, the produced list will be sorted in ascending order.

You may find it useful to define additional helper functions.


In [0]:
## Binary Search Tree
##
class Node:
    def __init__(self, val):
        self.l_child = None
        self.r_child = None
        self.parent = None
        self.data = val
        self.node_list = [self]

    def insert(self, node):
        """inserts a node into a *non-empty* tree rooted at the node, returns
        the root"""

        self.node_list.append(node)

        if self.data > node.data:
            if self.l_child is None:
                self.l_child = node
                node.parent = self
            else:
                self.l_child.insert(node)
        else:
            if self.r_child is None:
                self.r_child = node
                node.parent = self
            else:
                self.r_child.insert(node)
        return self
    
    def minimum(self):
        node = self
        while node.l_child != None:
            node = node.l_child
        return node

    def search_data(self, value):
        """searches a *non-empty* tree rooted at the node for a node with
        data = value, returns the value if found, None otherwise"""
        node = self.search(value)
        if node:
            return node.data
        else:
            return None
        
    def to_string(self): 
        print('self.data', self.data)
        root=self
        if not root: 
            return 'Nil'
        else: 
            r = root.r_child.to_string() if root.r_child else 'Nil'
            l = root.l_child.to_string() if root.l_child else 'Nil'
        return 'Node(' + str(root.data) + ' L: ' + l + ' R: ' + r + ')'
    
    def search(self, value):
      '''
      searches a non-empty BST rooted at the node for a node with data=value, 
      returns the node if found, None otherwise
      '''
      current_node = self
      while(current_node != None and current_node.data != value):
        if(value < current_node.data):
          current_node = current_node.l_child
        else: 
          current_node = current_node.r_child

      return current_node

    def transplant(self, old_node, new_node):
      '''
      Replaced the subtree rooted at old_node by
      the subtree rooted at new_node.
      '''
      if(old_node == self):
        new_node = self
      
      elif (old_node == old_node.parent.l_child):
        old_node.parent.l_child = new_node

      else:
        old_node.parent.r_child = new_node

      if (new_node != None):
        new_node.parent = old_node.parent

    def delete(self, value):
      '''
      if a node with data = value is present in the tree rooted at Node, 
      deletes that node and returns the root.
      '''
      ## finding the node where data = value
      node = self.search(value)
      
      # if the node has no left child
      # replace it with its right child
      if (node.l_child == None):
        self.transplant(node, node.r_child)

      # if the node has no right child
      # replace it with its left child
      elif (node.r_child == None):
        self.transplant(node, node.l_child)

      ## if it has both children
      else:
        successor = node.r_child.minimum()
        
        # if the successor stays in the right subtree but
        # it is not the right child of the node,
        # then replace it with its own right child
        if (successor.parent != node):
          self.transplant(successor, successor.r_child)
          successor.r_child = node.r_child
          successor.r_child.parent = successor

        # replace the deleting node by its successor
        self.transplant(node, successor)
        successor.l_child = node.l_child
        successor.l_child.parent = successor

      return self

    def inorder(self):
      '''
      Returns a list of all data in the tree rooted at root produced 
      using an in order traversal. When correctly implemented on a BST, 
      the produced list will be sorted in ascending order
      '''
      current_node = self

      if (current_node.data == None):
        return []
      else: 
          r = current_node.r_child.inorder() if current_node.r_child else []
          l = current_node.l_child.inorder() if current_node.l_child else []
      return l + [current_node.data] + r

    #raise NotImplementedError()


## Question 4. Validating the BST python code

### Question 4a.

It is good practice to make the necessary tests in your code to ensure it produces the intended outputs. In the cells below, implement slow, but simple:
* inserts,
* searches,
* deletes.


In [0]:
def list_insert(lst, value):
    """inserts value into lst in sorted order"""
    
    lst.append(value)
    lst = sorted(lst)

    # raise NotImplementedError()


In [0]:
def list_delete(lst, value):
    """ deletes first instance of value from lst if it present"""
    
    lst.remove(value)
    
    # raise NotImplementedError()

In [0]:
def list_search(lst, value): 
    """ searches lst for value and returns value if present, None if it is not present"""
    
    if value in lst:
      return value
    else:
      return None

    # raise NotImplementedError()

### Question 4b. 
Run the testing code provided in the cell below to generate a sequence of random inserts, followed by a sequence of random deletes, and finally followed by a sequence of random searches. Apply this sequence to both your BST implementation and the sorted list implementation. Do the final results both match? Does this mean your code is free of bugs? Provide your answer to these questions in the cell below the Python-code cell.

In [0]:
bst = None # bst is a misnormer, this variable contains the Node that is the root of the BST of interest
lst = []

my_Nodes = [2, 3, 5, 1]
for x in [Node(_) for _ in my_Nodes]:
    print("###################")
    print('Inserting the following node: ', x.data)
    if not bst:
        bst = x
    else:
        bst = bst.insert(x)
    print(bst.to_string())

###################
Inserting the following node:  2
self.data 2
Node(2 L: Nil R: Nil)
###################
Inserting the following node:  3
self.data 2
self.data 3
Node(2 L: Nil R: Node(3 L: Nil R: Nil))
###################
Inserting the following node:  5
self.data 2
self.data 3
self.data 5
Node(2 L: Nil R: Node(3 L: Nil R: Node(5 L: Nil R: Nil)))
###################
Inserting the following node:  1
self.data 2
self.data 3
self.data 5
self.data 1
Node(2 L: Node(1 L: Nil R: Nil) R: Node(3 L: Nil R: Node(5 L: Nil R: Nil)))


In [0]:
### Testing Code
###

import random

bst = None # bst is a misnormer, this variable contains the Node that is the root of the BST of interest
lst = []

# sampling with replacement
for x in [Node(random.randint(0,10)) for _ in range(10)]:
    print('Inserting the following node: ', x.data)
    if not bst:
        bst = x
    else:
        bst.insert(x)
    list_insert(lst, x.data)

print('In order traversal: ', bst.inorder())
print('The non-order list:', lst)

for x in [random.randint(0,10) for _ in range(4)]:
    print('Look for the following node: ', x)
    if bst.search_data(x):
        print(' Node exists, hence removing the Node...')
        bst = bst.delete(x)
    else:
        print(' Node does not exist and cannot be removed!')
    if list_search(lst, x):
        print(' Element exists in list, hence removing it...!')
        list_delete(lst,x)
    else:
        print(' Element does not exist in list, and cannot be removed!')

print('In order traversal of final BST: ', bst.inorder())
print('Final list: ', lst)







Inserting the following node:  5
Inserting the following node:  10
Inserting the following node:  7
Inserting the following node:  0
Inserting the following node:  1
Inserting the following node:  10
Inserting the following node:  6
Inserting the following node:  2
Inserting the following node:  0
Inserting the following node:  8
In order traversal:  [0, 0, 1, 2, 5, 6, 7, 8, 10, 10]
The non-order list: [5, 10, 7, 0, 1, 10, 6, 2, 0, 8]
Look for the following node:  4
 Node does not exist and cannot be removed!
 Element does not exist in list, and cannot be removed!
Look for the following node:  7
 Node exists, hence removing the Node...
 Element exists in list, hence removing it...!
Look for the following node:  2
 Node exists, hence removing the Node...
 Element exists in list, hence removing it...!
Look for the following node:  1
 Node exists, hence removing the Node...
 Element exists in list, hence removing it...!
In order traversal of final BST:  [0, 0, 5, 6, 8, 10, 10]
Final list:

In [0]:
bst = None # bst is a misnormer, # variable contains the Node that is the root of the BST of interest
lst = []

# sampling with replacement
for x in [Node(random.randint(0,10)) for _ in range(10)]:
    print('Inserting the following node: ', x.data)
    if not bst:
        bst = x
    else:
        bst.insert(x)
    list_insert(lst, x.data)
#print(bst.to_string())
print('In order traversal: ', bst.inorder())


for x in [random.randint(0,10) for _ in range(4)]:
    print('Look for the following node: ', x)
    if bst.search_data(x):
        print(' Node exists, hence removing the Node...')
        bst = bst.delete(x)
    else:
        print(' Node does not exist and cannot be removed!')
    if list_search(lst, x):
        print(' Element exists in list, hence removing it...!')
        list_delete(lst,x)
    else:
        print(' Element does not exist in list, and cannot be removed!')
#print(bst.to_string())


print('In order traversal of final BST: ', bst.inorder())
print('Final list: ', lst)



Inserting the following node:  5
Inserting the following node:  10
Inserting the following node:  5
Inserting the following node:  10
Inserting the following node:  9
Inserting the following node:  9
Inserting the following node:  10
Inserting the following node:  7
Inserting the following node:  6
Inserting the following node:  0
In order traversal:  [0, 5, 5, 6, 7, 9, 9, 10, 10, 10]
Look for the following node:  5
 Node exists, hence removing the Node...
 Element exists in list, hence removing it...!
Look for the following node:  0
 Node does not exist and cannot be removed!
 Element does not exist in list, and cannot be removed!
Look for the following node:  8
 Node does not exist and cannot be removed!
 Element does not exist in list, and cannot be removed!
Look for the following node:  9
 Node exists, hence removing the Node...
 Element exists in list, hence removing it...!
In order traversal of final BST:  [0, 5, 6, 7, 9, 10, 10, 10]
Final list:  [10, 5, 10, 9, 10, 7, 6, 0]


YOUR ANSWER HERE